main.rb

```diff
@@ -0,0 +1,213 @@
+# frozen_string_literal: true
+
+module Enumerable
+  def my_each
+    return to_enum(:my_each) unless block_given?
+
+    arr = self
+    counter = 0
+    loop do
+      yield arr.to_a[counter]
+      counter += 1
+      break if counter == arr.to_a.length
+    end
+    self
+  end
+
+  def my_each_with_index
+    return to_enum(:my_each) unless block_given?
+
+    arr = self
+    counter = 0
+    loop do
+      yield(arr.to_a[counter], counter)
+      counter += 1
+      break if counter == arr.to_a.length
+    end
+    self
+  end
+
+  def my_select
+    return to_enum(:my_select) unless block_given?
+
+    arr = to_a
+    counter = 0
+    selected_elts = []
+    until counter == arr.length
+      selected_elts << arr[counter] if yield(arr[counter])
+      counter += 1
+    end
+    selected_elts
+  end
+
+  def my_all?(arg = nil)
+    if block_given?
+      my_each do |elt|
+        return false unless yield(elt) == true
+      end
+      true
+    elsif arg.nil?
+      my_each { |elt| return false if elt.nil? || elt == false }
+    elsif !arg.nil? && (arg.is_a? Class)
+      my_each do |elt|
+        return true unless elt.instance_of?(arg.class)
+      end
+      false
+    elsif !arg.nil? && arg.instance_of?(Regexp)
+      my_each { |elt| return false unless elt.match(arg) }
+      true
+    end
+  end
+
+  def my_any?(arg = nil)
+    if block_given?
+      my_each { |elt| return true if yield(elt) }
+      return false
+    end
+    arg.nil? ? arg.class.to_s : my_any? { |elt| elt }
+
+    if arg.class.to_s == 'Class'
+      my_each { |elt| return true if elt.is_a? arg }
+    elsif arg.class.to_s == 'Regexp'
+      my_each { |elt| return true if elt =~ arg }
+    elsif arg.nil?
+      my_each { |elt| return true if elt }
+    else
+      my_each { |elt| return true if elt == arg }
+    end
+    false
+  end
+
+  def my_none?(arg = nil, &block)
+    !my_any?(arg, &block)
+  end
+
+  def my_count(number = nil, &block)
+    arr = to_a
+    return arr.length unless block_given? || number
+
+    return arr.my_select { |elt| elt == number }.length if number
+
+    arr.my_select(&block).length
+  end
+
+  def my_map(proc = nil)
+    return to_enum(:my_map) unless block_given? || proc.nil?
+
+    arr = self
+    new_arr = []
+    if proc.nil?
+      arr.my_each { |elt| new_arr << yield(elt) }
+    else
+      arr.to_a.my_each { |elt| new_arr << proc.call(elt) }
+    end
+    new_arr
+  end
```

Comment:
- [ ] `my_map` is not working as the original `map` method, also is not passing all test cases.

### `map` is an Enumerable that has these characteristics:
- Returns an Enumerator if no block is given.
```
[1, 2, 3].map
# #<Enumerator: [1, 2, 3]:map>
```

[Enumerable#map](https://ruby-doc.org/core-2.7.2/Enumerable.html#method-i-map)